pkg/stream/batch_consumer.go

```diff
@@ -12,53 +11,101 @@ import (
 type BatchConsumerCallback interface {
-	Boot(config cfg.Config, logger mon.Logger)
-	Consume(ctx context.Context, msg []*Message) error
+	Boot(cfg.Config, mon.Logger) error
+	Consume(context.Context, []*Message) ([]*Message, error)
+}
+
+type BatchConsumerSettings struct {
+	IdleTimeout time.Duration
+	BatchSize   int
 }
 
-type baseBatchConsumer struct {
+type batchConsumer struct {
 	kernel.EssentialModule
+	ConsumerAcknowledge
 
 	logger mon.Logger
 	tracer tracing.Tracer
+	metric mon.MetricWriter
 
-	input  Input
 	cfn    coffin.Coffin
 	ticker *time.Ticker
 
 	name     string
 	callback BatchConsumerCallback
+	settings BatchConsumerSettings
 
-	m         sync.Mutex
-	processed int
-	batchSize int
-	batch     []*Message
-	force     bool
+	lck   sync.Mutex
+	batch []*Message
 }
 
-func (c *baseBatchConsumer) Boot(config cfg.Config, logger mon.Logger) error {
-	c.logger = logger
-	c.tracer = tracing.NewAwsTracer(config)
-	c.cfn = coffin.New()
+func NewBatchConsumer(callback BatchConsumerCallback) Consumer {
+	return &batchConsumer{
+		cfn:      coffin.New(),
+		callback: callback,
+	}
+}
+
+func NewBatchConsumerWithInterfaces(callback BatchConsumerCallback, logger mon.Logger, tracer tracing.Tracer, input Input, metricWriter mon.MetricWriter, settings BatchConsumerSettings) Consumer {
+	c := NewBatchConsumer(callback).(*batchConsumer)
+
+	c.bootWithInterfaces(logger, tracer, input, metricWriter, settings)
+
+	return c
+}
+
+func (c *batchConsumer) Boot(config cfg.Config, logger mon.Logger) error {
+	err := c.callback.Boot(config, logger)
+
+	if err != nil {
+		return err
+	}
+
+	tracer := tracing.NewAwsTracer(config)
+
+	defaultMetrics := getDefaultConsumerMetrics()
+	mw := mon.NewMetricDaemonWriter(defaultMetrics...)
+
+	inputName := config.GetString("consumer_input")
+	input := NewConfigurableInput(config, logger, inputName)
 
 	idleTimeout := config.GetDuration("consumer_idle_timeout")
-	c.ticker = time.NewTicker(idleTimeout * time.Second)
+	batchSize := config.GetInt("consumer_batch_size")
 
-	c.batchSize = config.GetInt("consumer_batch_size")
-	c.batch = make([]*Message, 0, c.batchSize)
+	settings := BatchConsumerSettings{
+		IdleTimeout: idleTimeout,
+		BatchSize:   batchSize,
+	}
 
-	c.callback.Boot(config, logger)
+	c.bootWithInterfaces(logger, tracer, input, mw, settings)
 
 	return nil
 }
 
-func (c *baseBatchConsumer) Run(ctx context.Context) error {
+func (c *batchConsumer) bootWithInterfaces(logger mon.Logger, tracer tracing.Tracer, input Input, metricWriter mon.MetricWriter, settings BatchConsumerSettings) {
+	c.logger = logger
+	c.tracer = tracer
+	c.input = input
+	c.metric = metricWriter
+	c.settings = settings
+	c.ticker = time.NewTicker(settings.IdleTimeout)
+	c.batch = make([]*Message, 0, c.settings.BatchSize)
+}
+
+func (c *batchConsumer) Run(ctx context.Context) error {
 	defer c.logger.Info("leaving consumer ", c.name)
+	defer c.ticker.Stop()
+	defer c.consumeBatch(context.Background())
```

Comment:
This is more like a `processBatch`, right?

Reply:
What's the reason to use the background context? I consider you did this because aws services cancel the request when the context gets cancelled? In that case, i think it would be ok to use the supplied context of the `Run` function

Reply:
In this case, I want to make sure that the last batch is processed properly. When the context is already dead / canceled, any subsequent calls using it would fail. Hence, a new context.